File: conformance/reference-js/upload-execution.js
'use strict';

function failure(code, fileIndex, requests = []) {
  return { ok: false, code, fileIndex, requests, fieldValue: null };
}

const PROTOCOL_RELATIVE_URL = /^\/(?!\/)[^\s\\]*$/;
const INVOCATION_ID = /^[\x21-\x7e]{1,200}$/;

function matchesAccept(file, accept) {
  if (accept === undefined) return true;
  const fileName = file.name.toLowerCase();
  const mime = file.type.toLowerCase();
  return accept.split(',').some(rawToken => {
    const token = rawToken.trim().toLowerCase();
    if (token === '') return false;
    if (token.startsWith('.')) return fileName.endsWith(token);
    if (token.endsWith('/*')) return mime.startsWith(`${token.slice(0, -1)}`);
    return mime === token;
  });
}

function requestFor(action, file, fileIndex, invocationId) {
  const request = {
    method: action.method === undefined ? 'POST' : action.method,
    url: action.url,
    part: {
      name: action.fieldName === undefined ? 'file' : action.fieldName,
      fileName: file.name,
      contentId: file.contentId,
    },
  };
  const retryPolicy = action.retryPolicy === undefined ? 'never' : action.retryPolicy;
  if (!['never', 'idempotent'].includes(retryPolicy)) return failure('INVALID_RETRY_POLICY', fileIndex);
  if (retryPolicy === 'idempotent') {
    if (typeof invocationId !== 'string' || !INVOCATION_ID.test(invocationId)) {
      return failure('MISSING_INVOCATION_ID', fileIndex);
    }
    request.headers = { 'Idempotency-Key': `${invocationId}:${fileIndex}` };
  }
  return request;
}

function responseValue(response) {
  if (response === null || typeof response !== 'object' || Array.isArray(response)) return null;
  if (typeof response.url === 'string' && response.url.length > 0) return response.url;
  if (typeof response.id === 'string' && response.id.length > 0) return response.id;
  return null;
}

function executeUpload(input) {
  const { action, files } = input;
  if (typeof action.url !== 'string' || !PROTOCOL_RELATIVE_URL.test(action.url)) {
    return failure('INVALID_PROTOCOL_URL', 0);
  }
  if (!action.multiple && files.length > 1) return failure('MULTIPLE_FILES_NOT_ALLOWED', 1);
  for (let index = 0; index < files.length; index += 1) {
    const file = files[index];
    if (action.maxSize !== undefined && file.size > action.maxSize) {
      return failure('FILE_TOO_LARGE', index);
    }
    if (!matchesAccept(file, action.accept)) return failure('UNSUPPORTED_FILE_TYPE', index);
  }

  const requests = [];
  const values = [];
  for (let index = 0; index < files.length; index += 1) {
    const request = requestFor(action, files[index], index, input.invocationId);
    if (request.ok === false) return failure(request.code, index, requests);
    requests.push(request);
    const result = input.results[index];
    if (!result || result.type !== 'success') return failure('UPLOAD_REQUEST_FAILED', index, requests);
    const value = responseValue(result.response);
    if (value === null) return failure('INVALID_UPLOAD_RESPONSE', index, requests);
    values.push(value);
  }

  return {
    ok: true,
    requests,
    fieldValue: action.multiple ? values : values[0],
  };
}

module.exports = { executeUpload };

File: conformance/reference-python/upload_execution.py
import re


PROTOCOL_RELATIVE_URL = re.compile(r"^/(?!/)[^\s\\]*$")
INVOCATION_ID = re.compile(r"^[\x21-\x7e]{1,200}$")


def failure(code, file_index, requests=None):
    return {
        "ok": False,
        "code": code,
        "fileIndex": file_index,
        "requests": requests or [],
        "fieldValue": None,
    }


def matches_accept(file_value, accept):
    if accept is None:
        return True
    file_name = file_value["name"].lower()
    mime = file_value["type"].lower()
    for raw_token in accept.split(","):
        token = raw_token.strip().lower()
        if token == "":
            continue
        if token.startswith(".") and file_name.endswith(token):
            return True
        if token.endswith("/*") and mime.startswith(token[:-1]):
            return True
        if mime == token:
            return True
    return False


def request_for(action, file_value, file_index, invocation_id):
    request = {
        "method": action.get("method", "POST"),
        "url": action["url"],
        "part": {
            "name": action.get("fieldName", "file"),
            "fileName": file_value["name"],
            "contentId": file_value["contentId"],
        },
    }
    retry_policy = action.get("retryPolicy", "never")
    if retry_policy not in ("never", "idempotent"):
        return failure("INVALID_RETRY_POLICY", file_index)
    if retry_policy == "idempotent":
        if not isinstance(invocation_id, str) or INVOCATION_ID.fullmatch(invocation_id) is None:
            return failure("MISSING_INVOCATION_ID", file_index)
        request["headers"] = {"Idempotency-Key": f"{invocation_id}:{file_index}"}
    return request


def response_value(response):
    if not isinstance(response, dict):
        return None
    if isinstance(response.get("url"), str) and len(response["url"]) > 0:
        return response["url"]
    if isinstance(response.get("id"), str) and len(response["id"]) > 0:
        return response["id"]
    return None


def execute_upload(input_value):
    action = input_value["action"]
    files = input_value["files"]
    if not isinstance(action.get("url"), str) or PROTOCOL_RELATIVE_URL.fullmatch(action["url"]) is None:
        return failure("INVALID_PROTOCOL_URL", 0)
    if not action.get("multiple", False) and len(files) > 1:
        return failure("MULTIPLE_FILES_NOT_ALLOWED", 1)
    for index, file_value in enumerate(files):
        if "maxSize" in action and file_value["size"] > action["maxSize"]:
            return failure("FILE_TOO_LARGE", index)
        if not matches_accept(file_value, action.get("accept")):
            return failure("UNSUPPORTED_FILE_TYPE", index)

    requests = []
    values = []
    for index, file_value in enumerate(files):
        request = request_for(action, file_value, index, input_value.get("invocationId"))
        if request.get("ok") is False:
            return failure(request["code"], index, requests)
        requests.append(request)
        result = input_value["results"][index] if index < len(input_value["results"]) else None
        if result is None or result.get("type") != "success":
            return failure("UPLOAD_REQUEST_FAILED", index, requests)
        value = response_value(result.get("response"))
        if value is None:
            return failure("INVALID_UPLOAD_RESPONSE", index, requests)
        values.append(value)

    return {
        "ok": True,
        "requests": requests,
        "fieldValue": values if action.get("multiple", False) else values[0],
    }
